File: ingestion/transform/slugs.py
"""Slug generation — deterministic and stable across ingest re-runs."""

import re
import unicodedata


def slugify(text: str) -> str:
    """Lowercase, ASCII-safe, hyphen-separated slug."""
    text = unicodedata.normalize("NFKD", text)
    text = text.encode("ascii", "ignore").decode("ascii")
    text = text.lower()
    text = re.sub(r"[^a-z0-9]+", "-", text)
    return text.strip("-")


def driver_slug(forename: str, surname: str) -> str:
    return slugify(f"{forename} {surname}")


def race_slug(year: int | str, name: str) -> str:
    return slugify(f"{year} {name}")


def deduplicate(slug: str, existing: set[str]) -> str:
    """Append a numeric suffix until the slug is unique."""
    if slug not in existing:
        return slug
    n = 2
    while f"{slug}-{n}" in existing:
        n += 1
    return f"{slug}-{n}"

File: ingest/slugs.ts
export function slugify(text: string): string {
  // Fold to ASCII (NFKD decompose, strip combining marks, keep only ASCII letters/digits)
  const ascii = text
    .normalize('NFKD')
    .replace(/[̀-ͯ]/g, '')  // strip combining diacritical marks
    .replace(/[^\x00-\x7F]/g, '');   // strip remaining non-ASCII
  return ascii
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '');
}

export function driverSlug(forename: string, surname: string): string {
  return slugify(`${forename} ${surname}`);
}

export function raceSlug(year: number | string, name: string): string {
  return slugify(`${year} ${name}`);
}

export function deduplicate(slug: string, existing: Set<string>): string {
  if (!existing.has(slug)) return slug;
  let n = 2;
  while (existing.has(`${slug}-${n}`)) n++;
  return `${slug}-${n}`;
}
